adr/context.py

```diff
@@ -0,0 +1,287 @@
+        Get corresponding value for token
+        In this function, we return the same token value without using context
+        :param token: token object
+        :param pc: parse context
+        :return: same token value
+        """
+        self.contexts.add(token.value)
+        return token.value
+
+    def parse(self, expression):
+        """
+        Parse current expression, return contexts inside this expression
+        :param str expression: expression
+        :return: set of contexts
+        """
+        self.contexts.clear()
+        super(ZeroContextEvaluator, self).parse(expression)
+        return self.contexts
+
+
+G_ZEROCONTEXT_EVALUATOR = ZeroContextEvaluator()
+
+
+def extract_context_names(query):
+    """
+    Extract contexts list from query object
+    This is to mimic json-e render() function
+    Given using only $eval in query
+    :param dict query: query object to extract context
+    :return: set of context names
+    """
+    contexts = set()
+    if isinstance(query, dict):
+        for k, v in query.items():
+            if k == "$eval":
+                contexts.update(G_ZEROCONTEXT_EVALUATOR.parse(v))
+            else:
+                contexts.update(extract_context_names(v))
+    elif isinstance(query, list):
+        for v in query:
+            contexts.update(extract_context_names(v))
+
+    return contexts
+
+
+def _get_destination(definition):
+    """
+    Extract destination name from definition
+    :param list definition: definition of context, in form of list or str
+    :return:
+    """
+    if not isinstance(definition, list):
+        return None
+    if len(definition) < 2:
+        raise TypeError("Invalid definition {}".format(definition))
+    arg_def = definition[1]
+    keys = definition[0]
+    # If dest is defined, use it
+    if 'dest' in arg_def:
+        return arg_def['dest']
+    # Otherwise use the long form and/or short form of arg key name
+    val = None
+    for key in keys:
+        # If found long form arg, found it!
+        if key.startswith('--'):
+            val = key[2:]
+            break
+        # If found short form, store value but keep searching (for long form)
+        elif key.startswith('-'):
+            val = key[1:]
+    return val
+
+
+def _get_context_definition(context_name, definitions):
+    for definition in definitions:
+        if _get_destination(definition) == context_name:
+            return definition
+
+
+def get_context_definitions(definitions):
+    """
+    Build full context definitions from list of contexts with pre-defined definitions
+    :param list definitions: list of context definition, in which each definition is an array
+    :return:
+    """
+    result = []
+    destinations = set()
+    for name in definitions:
+        if isinstance(name, list):
+            definition = name
+        elif not isinstance(name, str):
+            raise TypeError("Only accept str/list, but found {}".format(type(name)))
+        else:
```

Comment:
nit: swap the elif/else so that the raise happens in the else clause. It's slightly more consistent.

Reply:
Thanks, updated. Please refer to https://github.com/mozilla/active-data-recipes/pull/85